src/backend/distributed/executor/distributed_intermediate_results.c

```diff
@@ -318,3 +379,227 @@ ExecuteSelectTasksIntoTupleStore(List *taskList, TupleDesc resultDescriptor)
+				lappend(fragmentListTransfer->fragmentList, fragment);
+		}
+	}
+
+	List *fragmentListTransfers = NIL;
+	NodeToNodeFragmentsTransfer *transfer = NULL;
+	HASH_SEQ_STATUS hashSeqStatus;
+
+	hash_seq_init(&hashSeqStatus, transferHash);
+
+	while ((transfer = hash_seq_search(&hashSeqStatus)) != NULL)
+	{
+		fragmentListTransfers = lappend(fragmentListTransfers, transfer);
+	}
+
+	return fragmentListTransfers;
+}
+
+
+/*
+ * FragmentTransferTaskList returns a list of tasks which performs the given list of
+ * transfers. Each of the transfers are done by a SQL call to fetch_intermediate_results.
+ * See QueryStringForFragmentsTransfer for how the query is constructed.
+ */
+static List *
+FragmentTransferTaskList(List *fragmentListTransfers)
+{
+	List *fetchTaskList = NIL;
+	ListCell *transferCell = NULL;
+
+	foreach(transferCell, fragmentListTransfers)
+	{
+		NodeToNodeFragmentsTransfer *fragmentsTransfer = lfirst(transferCell);
+
+		int targetNodeId = fragmentsTransfer->nodes.targetNodeId;
+
+		/* these should have already been pruned away in ColocationTransfers */
+		Assert(targetNodeId != fragmentsTransfer->nodes.sourceNodeId);
+
+		WorkerNode *workerNode = LookupNodeByNodeId(targetNodeId);
+
+		ShardPlacement *targetPlacement = CitusMakeNode(ShardPlacement);
+		targetPlacement->nodeName = workerNode->workerName;
+		targetPlacement->nodePort = workerNode->workerPort;
+		targetPlacement->groupId = workerNode->groupId;
+
+		Task *task = CitusMakeNode(Task);
+		task->taskType = SELECT_TASK;
+		task->queryString = QueryStringForFragmentsTransfer(fragmentsTransfer);
+		task->taskPlacementList = list_make1(targetPlacement);
+
+		fetchTaskList = lappend(fetchTaskList, task);
+	}
+
+	return fetchTaskList;
+}
+
+
+/*
+ * QueryStringForFragmentsTransfer returns a query which fetches distributed
+ * result fragments from source node to target node. See the structure of
+ * NodeToNodeFragmentsTransfer for details of how these are decided.
+ */
+static char *
+QueryStringForFragmentsTransfer(NodeToNodeFragmentsTransfer *fragmentsTransfer)
+{
+	ListCell *fragmentCell = NULL;
+	StringInfo queryString = makeStringInfo();
+	StringInfo fragmentNamesArrayString = makeStringInfo();
+	int fragmentCount = 0;
+	NodePair *nodePair = &fragmentsTransfer->nodes;
+	WorkerNode *sourceNode = LookupNodeByNodeId(nodePair->sourceNodeId);
+
+	appendStringInfoString(fragmentNamesArrayString, "ARRAY[");
+
+	foreach(fragmentCell, fragmentsTransfer->fragmentList)
+	{
+		DistributedResultFragment *fragment = lfirst(fragmentCell);
+		char *fragmentName = fragment->resultId;
+
+		if (fragmentCount > 0)
+		{
+			appendStringInfoString(fragmentNamesArrayString, ",");
+		}
+
+		appendStringInfoString(fragmentNamesArrayString,
+							   quote_literal_cstr(fragmentName));
+
+		fragmentCount++;
+	}
+
+	appendStringInfoString(fragmentNamesArrayString, "]::text[]");
+
+	appendStringInfo(queryString,
+					 "SELECT fetch_intermediate_results(%s,%s,%d) bytes",
```

Comment:
I think doing `SELECT * FROM fetch_intermediate_results` is a little nicer. If we add multiple return values in the future and some workers are ahead of the coordinator, then it won't break this query.

Reply:
done